Refactor cloud-linux page (#40636)
gravitational/teleport: docs/pages/upgrading/cloud-kubernetes.mdx

---

This guide explains how to upgrade Teleport Cloud agents running on Kubernetes.
To keep agents up to date, you can either perform:
- [Automatic updates](#automatic-updates) (recommended)

> - [Manual updates](#manual-updates)

Reviewer: If we recommend automatic updates, I would make it easier for users to enable automatic updates and more difficult to enable manual updates. We should frame this guide around automatic updates and describe manual updates as an optional users can take if automatic updates are not possible.

Author: We want to push users to use automatic updates, but there are still many users that are not ready or do not want to use on automatic updates. For these users, if we make it difficult to perform manual updates, I think they will be more likely to just not upgrade their agents/clusters.

I'd like to make the instructions for both methods easy to follow, but make it clear that automatic updates is the better method. And to highlight that if they opt out of automatic updates, they need to perform updates weekly, or at the very least for every major version release.

## Prerequisites

- Familiarity with the [Upgrading Compatibility Overview](./overview.mdx) guide,
  which describes the sequence in which to upgrade components of your cluster.

- (!docs/pages/includes/cloud/tctl-tsh-prerequisite.mdx!)

- (!docs/pages/includes/tctl.mdx!)

> ## Check for agent updates

Reviewer: We use the "Step n/d." format in H2s in how-to guides so readers know how far along they are. I would remove the "Automatic updates" H2 and turn the steps within that into "Step" headings.

> The latest compatible Teleport agent version is served by the Teleport Proxy. To
> identify what that version is, you can query the `webapi` endpoint.

Reviewer: ```suggestion
To identify the latest compatible Teleport agent version, you can query the `webapi` endpoint of the Teleport Proxy Service:

```
To make this more concise

```code
$ curl https://<Var name="teleport.example.com" />/webapi/automaticupgrades/channel/stable/cloud/version
v15.2.1
```

> To list the agents that are upgradable, use the `tctl inventory ls` command to list

Reviewer: > To list the agents that are upgradable, use the `tctl inventory ls` command to list
the inventory of connected agents.

I would condense the two "to" phrases here.

the inventory of connected agents. The agents are listed along with their current
version. Additional flags are available to filter by service `--services=node,kube,db,...`,

> or to filter by version `--older-than=v15.2.1`.

Reviewer: > Additional flags are available to filter by service `--services=node,kube,db,...`,
or to filter by version `--older-than=v15.2.1`.

I would expand this sentence and use conventional punctuation and sentence structure. Right now, this looks fairly jumbled.

```code
$ tctl inventory ls --older-than=v15.2.1
Server ID                            Hostname       Services       Version Upgrader
------------------------------------ -------------- -------------- ------- --------
00000000-0000-0000-0000-000000000000 teleport-agent Kube           v14.4.5 none
00000000-0000-0000-0000-000000000001 teleport-proxy Proxy          v15.2.0 none
00000000-0000-0000-0000-000000000002 teleport-auth  Auth,Discovery v15.2.0 none
...
```
Note that the `inventory ls` command will also list `teleport-auth` and `teleport-proxy`
services. These services are managed by the Teleport Cloud team and updates will
be performed automatically.

## Automatic updates

Agents enrolled in automatic updates will automatically be kept up to date with the
latest compatible version of Teleport. You can find more information regarding the

> automatic updates architecture in the [Agent Update Management](../architecture/agent-update-management.mdx)

Reviewer: We should move this link to the end of the page per the [style guide](https://github.com/gravitational/docs/blob/main/docs-contributors/style-guide.md#using-links-admonishments-and-details-in-how-to-guides).

page.

> Automatic updates is not enabled by default for Helm releases. The `teleport-kube-agent`

Reviewer: ```suggestion
Automatic updates are not enabled by default for Helm releases. The `teleport-kube-agent`
```

Helm chart values must be configured to enable automatic updates. For more details

> see the `teleport-kube-agent` [chart reference](../reference/helm-reference/teleport-kube-agent.mdx#updater).

Reviewer: Since this guide is telling the reader how to enable automatic updates, would it make sense to remove this sentence? If we do want to link the reader to the `teleport-kube-agent` page, we should do so at the end of the guide.

This section assumes that the name of your `teleport-kube-agent` release is
`teleport-agent`, and that you have installed it in the `teleport` namespace.

(!docs/pages/kubernetes-access/helm/includes/helm-repo-add.mdx!)

Add the following chart values to your existing agent `values.yaml` to enable
the automatic updater:

```yaml
updater:
    enabled: true
```

Upgrade the Helm chart release with the new values by running `helm upgrade`:

```code
$  helm -n <Var name="teleport" /> upgrade <Var name="teleport-agent" /> teleport/teleport-kube-agent  \
--values=values.yaml \
--version=<Var name="(=cloud.version=)" />
```

## Manual updates

[Automatic updates](#automatic-updates) is the recommended update method for Teleport
agents. If you choose to opt out of automatic updates, you will need to be responsible
for keeping your Teleport agents up to date with the version of Teleport Cloud.

(!docs/pages/kubernetes-access/helm/includes/helm-repo-add.mdx!)

Upgrade the Helm release:

```code
$ helm -n <Var name="teleport" /> upgrade <Var name="teleport-agent" /> teleport/teleport-kube-agent \
--values=values.yaml \
--version=<Var name="(=cloud.version=)" />
```

> ## FAQ

Reviewer: I would rename this H2 and move the "Manual updates" section within it, since this H2 includes instructions for nonstandard situations. Maybe we could call this section "Edge cases" or similar?
